inc/Integrations/SalesforceB2C/Auth/SalesforceB2CAuth.php

```diff
@@ -0,0 +1,239 @@
+		if ( null !== $saved_refresh_token ) {
+			$access_token = self::get_token_using_refresh_token( $saved_refresh_token, $client_id, $client_secret, $endpoint, $organization_id );
+		}
+
+		if ( null !== $access_token ) {
+			return $access_token;
+		}
+
+		$access_token = self::get_token_using_client_credentials( $client_id, $client_secret, $endpoint, $organization_id );
+		return $access_token;
+	}
+
+	// Access token request using top-level credentials
+
+	public static function get_token_using_client_credentials(
+		string $client_id,
+		string $client_secret,
+		string $endpoint,
+		string $organization_id,
+	): WP_Error|string {
+		$client_auth_url = sprintf( '%s/shopper/auth/v1/organizations/%s/oauth2/token', $endpoint, $organization_id );
+		$client_credentials = base64_encode( sprintf( '%s:%s', $client_id, $client_secret ) );
+
+		$client_auth_response = wp_remote_post( $client_auth_url, [
+			'body' => [
+				'grant_type' => 'client_credentials',
+				'channel_id' => 'RefArch',
+			],
+			'headers' => [
+				'Content-Type' => 'application/x-www-form-urlencoded',
+				'Authorization' => 'Basic ' . $client_credentials,
+			],
+		]);
+
+		if ( is_wp_error( $client_auth_response ) ) {
+			return new WP_Error(
+				'salesforce_b2c_auth_error_client_credentials',
+				__( 'Failed to retrieve access token from client credentials', 'remote-data-blocks' )
+			);
+		}
+
+		$response_code = wp_remote_retrieve_response_code( $client_auth_response );
+		$response_body = wp_remote_retrieve_body( $client_auth_response );
+		$response_data = json_decode( $response_body, true );
+
+		if ( 400 === $response_code ) {
+			return new WP_Error(
+				'salesforce_b2c_auth_error_client_credentials',
+				/* translators: %s: Technical error message from API containing failure reason */
+				sprintf( __( 'Failed to retrieve access token from client credentials: "%s"', 'remote-data-blocks' ), $response_data['message'] )
+			);
+		}
+
+		$access_token = $response_data['access_token'];
+		$access_token_expires_in = $response_data['expires_in'];
+		self::save_access_token( $access_token, $access_token_expires_in, $organization_id, $client_id );
+
+		$refresh_token = $response_data['refresh_token'];
+		$refresh_token_expires_in = $response_data['refresh_token_expires_in'];
+		self::save_refresh_token( $refresh_token, $refresh_token_expires_in, $organization_id, $client_id );
+
+		return $access_token;
+	}
+
+	// Access token request using refresh token
+
+	public static function get_token_using_refresh_token(
+		string $refresh_token,
+		string $client_id,
+		string $client_secret,
+		string $endpoint,
+		string $organization_id,
+	): ?string {
+		$client_auth_url = sprintf( '%s/shopper/auth/v1/organizations/%s/oauth2/token', $endpoint, $organization_id );
+
+		// Even though we're using a refresh token, authentication is still required to receive a new secret
+		$client_credentials = base64_encode( sprintf( '%s:%s', $client_id, $client_secret ) );
```

Comment:
I think this is a bit silly - in order to refresh our auth token using a refresh token, we still need to pass in our client ID and secret. From what I can tell, this is due to the OAuth spec. See "Client Authentication" on [Refreshing Access Tokens](https://www.oauth.com/oauth2-servers/access-tokens/refreshing-access-tokens/):

> If the client was issued a secret, then the client must authenticate this request.

Because we receive a secret in the initial grant, we still need it again here.

It makes me wonder - is it really necessary to save and reuse the refresh token? We'll still sending client credentials every 30 minutes anyway, why not just use the `client_credentials` grant every time and remove this function entirely?

Reply:
I believe the primary reason to use the refresh token is because you're generating revokable tokens on the Authorization Server. So imagine we used the client credentials to get new tokens every time instead of just refreshing the one... and then the user goes in to manage their issued tokens and revoke some... and they see a list of 10,000 tokens because we've created a new one on every refresh.